src/main/java/org/opensearch/ad/ratelimit/RateLimitedQueue.java

```diff
@@ -0,0 +1,474 @@
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.ad.AnomalyDetectorPlugin;
+import org.opensearch.ad.ExpiringState;
+import org.opensearch.ad.MaintenanceState;
+import org.opensearch.ad.NodeStateManager;
+import org.opensearch.ad.breaker.ADCircuitBreakerService;
+import org.opensearch.ad.common.exception.AnomalyDetectionException;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.threadpool.ThreadPoolStats;
+
+/**
+ * HCAD can bombard ES with “thundering herd” traffic, in which many entities
+ * make requests that need similar ES reads/writes at approximately the same
+ * time. To remedy this issue we queue the requests and ensure that only a
+ * limited set of requests are out for ES reads/writes.
+ *
+ * @param <RequestType> Individual request type that is a subtype of ADRequest
+ */
+public abstract class RateLimitedQueue<RequestType extends QueuedRequest> implements MaintenanceState {
+    /**
+     * Each request is associated with a segment. That is, a queue consists of segments.
+     * Segments have their corresponding priorities: HIGH, MEDIUM, and LOW. An example
+     * of HIGH priority requests is anomaly results with errors or its anomaly grade
+     * larger than zero. An example of MEDIUM priority requests is a cold start request
+     * for an entity. An example of LOW priority requests is checkpoint write requests
+     * for a cold entity. LOW priority requests have the slightest chance to be selected
+     * to be executed. MEDIUM and HIGH priority requests have higher stakes. LOW priority
+     * requests have higher chances of being deleted when the size of the queue reaches
+     * beyond a limit compared to MEDIUM/HIGH priority requests.
+     *
+     */
+    class Segment implements ExpiringState {
+        // last access time
+        private Instant lastAccessTime;
+        // data structure to hold requests
+        private BlockingQueue<RequestType> content;
+
+        Segment() {
+            this.lastAccessTime = clock.instant();
+            this.content = new LinkedBlockingQueue<RequestType>();
+        }
+
+        @Override
+        public boolean expired(Duration stateTtl) {
+            return expired(lastAccessTime, stateTtl, clock.instant());
+        }
+
+        public void put(RequestType request) throws InterruptedException {
+            this.content.put(request);
+        }
+
+        public int size() {
+            return this.content.size();
+        }
+
+        public void setLastAccessTime(Instant lastAccessTime) {
+            this.lastAccessTime = lastAccessTime;
```

Comment:
This does not have to be precise, just a signal for unused old segment that can be removed.  It is fine if we have race condition.  Don't want to synchronize the access as this could penalize performance. added comments.